From 9848d01231be062c3bc18775a28d37aec71651cc Mon Sep 17 00:00:00 2001
From: Icenowy Zheng <uwu@icenowy.me>
Date: Sat, 19 Jul 2025 14:05:13 +0800
Subject: [PATCH 390/400] drm/etnaviv: set the 32-bit coherent DMA mask later

The etnaviv driver wants to use 40-bit DMA mask and 32-bit coherent
allocation mask, to be able to access memory up to 40-bit, but still
limit page tables to lower 32-bit.

However, the call to of_dma_configure() next to setting the masks ruined
everything -- of_dma_configure will use the coherent allocation mask to
clip the DMA accessibility mask, leads to a setup with both masks as
32-bit.

In this situation PRIME imports stop to work because of SWIOTLB getting
activated.

Set a 40-bit coherent mask before of_dma_configure() and then the real
32-bit mask after it, to prevent main DMA mask being clipped.

Signed-off-by: Icenowy Zheng <uwu@icenowy.me>
---
 drivers/gpu/drm/etnaviv/etnaviv_drv.c | 13 +++++++++++--
 1 file changed, 11 insertions(+), 2 deletions(-)

diff --git a/drivers/gpu/drm/etnaviv/etnaviv_drv.c b/drivers/gpu/drm/etnaviv/etnaviv_drv.c
index 85d0695e94a5..39f13b512261 100644
--- a/drivers/gpu/drm/etnaviv/etnaviv_drv.c
+++ b/drivers/gpu/drm/etnaviv/etnaviv_drv.c
@@ -632,9 +632,11 @@ static int etnaviv_pdev_probe(struct platform_device *pdev)
 	 * To make things easy, we set the dma_coherent_mask to 32
 	 * bit to make sure we are allocating the command buffers and
 	 * TLBs in the lower 4 GiB address space.
+	 *
+	 * As coherent_mask may mislead of_dma_configure(), set it to
+	 * the same with main dma mask temporarily as a hack.
 	 */
-	if (dma_set_mask(&pdev->dev, DMA_BIT_MASK(40)) ||
-	    dma_set_coherent_mask(&pdev->dev, DMA_BIT_MASK(32))) {
+	if (dma_set_mask_and_coherent(&pdev->dev, DMA_BIT_MASK(40))) {
 		dev_dbg(&pdev->dev, "No suitable DMA available\n");
 		return -ENODEV;
 	}
@@ -647,6 +649,13 @@ static int etnaviv_pdev_probe(struct platform_device *pdev)
 	if (first_node)
 		of_dma_configure(&pdev->dev, first_node, true);
 
+	/*
+	 * Now as of_dma_configure() is done, we're now safe to set
+	 * coherent_dma_mask to the real value.
+	 */
+	if (dma_set_coherent_mask(&pdev->dev, DMA_BIT_MASK(32)))
+		return -ENODEV;
+
 	return component_master_add_with_match(dev, &etnaviv_master_ops, match);
 }
 
-- 
2.43.0

